### DISCUSSION TITLE
Refreshing `FutureProvider`'s async state

### DISCUSSION THREAD

**Discussion by Maldus512** [rrousselGit/riverpod#2108] [Q&A] [answered]
Hello everyone,
I am starting to use Riverpod for a new application and I'm trying to learn my way around providers.

In my situation I have a `StateNotifierProvider` whose `StateNotifier` contains the global state of my application. Some of that state can be fetched from a remote server and the corresponding operation is, of course, async.

In a page I'd like for example to show an information that can change on the server and thus should be refreshable. Long story short, my `FutureProvider` never goes back to the `loading` state after the first round. It always starts a refresh operation and the data is correctly returned, but I'd like to show a loading spinner every time the data is actually loading.

I've written a simplified example with the objective I'd like to achieve:

```
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

final stateProvider = StateProvider<int>((_) => 0);
final futureProvider = FutureProvider<int>((ref) async {
  await Future.delayed(const Duration(seconds: 1));
  ref.read(stateProvider.notifier).state += 1;
  return ref.watch(stateProvider.notifier).state;
});

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends ConsumerWidget {
  const MyApp({super.key});

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Demo"),
        ),
        body: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              const Text(
                'You have pushed the button this many times:',
              ),
              ref.watch(futureProvider).when(
                    data: (data) => Text("$data"),
                    error: (error, _) => Text("$error"),
                    loading: () => const Text("Loading"),
                  ),
            ],
          ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: () => _incrementCounter(ref),
          tooltip: 'Increment',
          child: const Icon(Icons.add),
        ), // This trailing comma makes auto-formatting nicer for build methods.
      ),
    );
  }

  void _incrementCounter(WidgetRef ref) {
    ref.invalidate(futureProvider);
  }
}
``` 

Basically I'd like the "Loading" text to appear every time the `FutureProvider` is loading; instead after the first time it just returns the cached value as ready data. Is this a bug or am I missing something?

**Comment by AhmedLSayed9** [comment, marked as answer]
It's a feature not a bug. You can add `skipLoadingOnRefresh: false` to `.when` to avoid skipping loading when doing ref.refresh/ref.invalidate.
